akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -122,6 +122,16 @@ private[http] object HttpServerBluePrint {
         completeStage()
       }
 
+      override def onUpstreamFinish(): Unit = super.onUpstreamFinish()
+      override def onUpstreamFailure(ex: Throwable): Unit = {
```

Comment:
Before the stage was just closed and the entity source potentially left hanging. There seemed to be some kind of behavior which actually prevented that but let's be rather be explicitly and propagate the failure if we are not able to send any more data to it.